### ISSUE TITLE
New Feature Suggestion: Split packages and custom mutation methods

### ISSUE THREAD

**Issue by christianalfoni** [christianalfoni/cerebral#61]
Hi guys!

I just got to say big thanks for fantastic feedback on Cerebral. The latest big contribution was the idea of adding outputs and type checking, by @a-s-o. This makes Cerebral able to understand even more about your flow and help you keep control of how signals run, especially on larger projects and with teams. Also hope the new "output" indication in the debugger is helpful :-)

There are two new great ideas I want to relay here:

# 1. Split Cerebral packages into two types
By: @gcanti

Now we have packages like `cerebral-react-immutable-store`, `cerebral-angular-immutable-store` and `cerebral-react-baobab`. What if we had: `cerebral-react`, `cerebral-angular`, `cerebral-immutable-store` and `cerebral-baobab`?

This would allow us to reuse packages and you would make a more conscious choice on choosing what view layer to use and what model layer to use.

This also leads into the next suggestion.

# 2. Custom state mutation methods
By: @pstoica

Baobab and Immutable-Store has very similar APIs, but Immutable-JS is quite different. When thinking of the previous example it would be more natural to make mutations pr. `model-layer-package`. So in the documentation of each model-layer package it would describe what mutation methods are available to you. That way Cerebral can take the full power of the model-layer and you actually work with the model-layer you choose, not Cerebral default mutations which can feel limiting if you want to use immutable-js for example.

So current packages will keep their current mutation methods, but when using `cerebral-immutable-js` you might use additional/different methods.

**UPDATE:** A revised suggestion is to make the implementation of creating mutation hooks very flexible, but by convention the official Cerebral Model Packages will share the same API (set, push, pop etc.). That allows others to create custom controllers with any state mutation methods for their own project or to share with others.

# Some syntax suggestions
```js
import Controller from 'cerebral';
import Model from 'cerebral-baobab';
import View from 'cerebral-react';

const state = {...};

const defaultInput = {...};

const controller = Controller({
  view: View,
  model: Model,
  state: state,
  defaultInput: defaultInput
});

export default controller;
```
So basically when the controller instantiates itself there has to be a common way to connect this stuff together. Needs some more thought, but just wanted to start somewhere.

What are the thoughts on this?

**Comment by Guria** [comment]
I think first point is must-have. Second also sounds reasonable.

**Comment by a-s-o** [comment]
Thanks for the mention Christian, but credit goes to you for being open to feedback and working hard on the implementation.

I agree with the first change. I am using baobab and mithril so I really just had to implement my own store hooks based on baobab-react. Separating packages will definitely help and also let the packages just focus on their main goals.

The second could also make cerebral more flexible but so far, I like the mutation hooks in cerebral. I also think it makes the system more robust as I am able to switch out the underlying store if need be without changing any mutations called in the actions. For example, switching to local-storage, immutable-store, etc. My suggestion is just to have a consistent api in the `Model` above

**Comment by christianalfoni** [comment]
Thanks for the input @a-s-o. You definitely make a valid point!

What we could do though is make the implementation flexible, but lock the API for the official packages. So if you create your own custom package you can use whatever mutation methods you want. BUT, the official cerebral-packages share a common mutation API to make it easy to change model-layer.

So @pstoica could build a custom immutable-js controller for the specific project or share on github, with mutation methods fitted for immutable-js. But the official `cerebral-immutable-js` project has the standard "set, push, pop" etc. methods.

**Comment by pstoica** [comment]
Hey all! Definitely wanted the first feature as soon as I copied over the React helpers. I did need a custom `shouldComponentUpdate` using `Immutable.is` since I decided to pass the immutable objects to components without serializing with `toJS` (useful when dealing with larger objects as props). So there could be some cross-cutting concerns between the model and view layers.

For the second, it could be good to have a consistent mutation API for the official packages and leave room for custom hooks. On the other hand, if we already select and receive data differently in the view layer, maybe it's okay for hooks to vary officially.

Immutable has its own gotchas that you should probably be familiar with if you choose it. For example:

* `Immutable.Map().set('a', {b: 1})` will have the plain, mutable object in 'a'. `merge` will actually turn it into an `Immutable.Map`. If example Cerebral code uses set with a plain object, I'm not sure if that code should be portable or if the adapter needs to call `fromJS` or something.
* You might also have different function names and signatures between Immutable methods and Cerebral methods. For example, Immutable's `setIn` already takes a path, which might make Cerebral's `set` confusing.
* Immutable has some mutators that take functions. These won't be serializable, but there should be a way to warn if you try to use them. However, `updateIn` is the most generic way to update a path, and there should be a way to make that serializable, e.g. `state.updateIn(['a', 'b'], 'pop')` would call `.pop()` on the item at that path.

Would typechecking be the proper solution for mutation hooks, just like in `output`? I'm starting to think Cerebral should focus on making mutations serializable/deterministic without worrying about standardizing them. If we do that, the Cerebral controller could easily passthrough the defined hooks to the Immutable methods.

I can try to come up with more Immutable vs standard Cerebral differences later if that would help more.

**Comment by christianalfoni** [comment]
Hello there!

Okay, so the new version of Cerebral is out with split VIEW/MODEL layer. Kinda cool to just swap out immutable-store with Baobab and it just works :-) I also tried to document the difference between the different model layers.

I am not too familiar with immutable js, but will check it out as soon as I can. Create a conventional implementation first with default mutation methods. But now it is possible to create any mutation methods you want, so it will be fun to explore how this will work out :-)

**Comment by christianalfoni** [comment]
Creating a new feature issue, closing this one
